junit-platform-engine/src/main/java/org/junit/platform/engine/TestDescriptor.java

```diff
@@ -159,9 +162,31 @@ default boolean isRoot() {
 
 	/**
 	 * Determine if this descriptor or any of its descendants describes a test.
+	 *
+	 * <p>This method is considered internal API, or in Java 9 syntax: {@code private}
 	 */
+	@API(Internal)
 	default boolean hasTests() {
-		return (isTest() || getChildren().stream().anyMatch(TestDescriptor::hasTests));
+		return isTest() || getChildren().stream().anyMatch(TestDescriptor::hasTests);
+	}
+
+	/**
+	 * Remove this descriptor from hierarchy unless it is a root descriptor or
+	 * itself is a test or any of its children describes a test.
+	 */
```

Comment:
This method actually opens up an interesting use case: An engine might override `prune` and do it differently or skip it altogether. I think that's actually a good idea! 👍 

However, I think the Javadoc should be more clear on this use case and should clarify that this is what the default implementation does.